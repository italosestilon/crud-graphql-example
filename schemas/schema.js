import {GraphQLSchema} from 'graphql';

import ArtistType from '../types/artist';
import Artist from '../models/artist';

import QueryType from '../queries/index';
import MutationType from '../mutations/index';

export const schema = new GraphQLSchema({
	query: QueryType,
	mutation: MutationType
});