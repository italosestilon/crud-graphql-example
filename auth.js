import jwt from "jsonwebtoken";
import config from "config";
import User from "./models/user";

export async function getUser(token) {
  if(!token) {
    return {user: null}
  };

  try {
    const decoded = jwt.verify(token.substring(4), config.jwtSecret);
    const user = await User.FindById(decoded.id);

    return {
      user
    };
  } catch (err) {
    console.error(err);
    return {
      user: null
    };
  }
}

export function generateToken(user) {
  return `JWT ${jwt.sign({ id: user.id }, config.jwtSecret)}`;
}
