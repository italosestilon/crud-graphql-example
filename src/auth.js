import jwt from "jsonwebtoken";
import User from "./models/user";

export async function getUser(token) {
  if (!token) {
    return { user: null };
  }

  try {
    const decoded = jwt.verify(token.substring(4), process.env.JWT_KEY);
    const user = await User.findById(decoded.id);

    return {
      user
    };
  } catch (err) {
    console.error(err);
    return {
      user: null
    };
  }
}

export function generateToken(user) {
  return `JWT ${jwt.sign({ id: user.id }, process.env.JWT_KEY)}`;
}
