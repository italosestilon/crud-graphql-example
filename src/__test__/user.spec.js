import { graphql } from "graphql";

import { schema } from "../schemas/schema";
import User from "../models/user";
import auth from "../auth";

import {
  connect,
  clearDatabase,
  disconnect,
  sanitizeObject,
  getContext
} from "../../test/helper";

beforeAll(connect);
afterEach(clearDatabase);
afterAll(disconnect);

describe("Graphql API", async () => {
  it("should register an user", async () => {
    const mutation = `
      mutation {
        registerWithEmail(input: {
        name : "Lana Del Rey",
        email: "delrey@email.com",
        password: "veryGoodPassword"
      }){
          token
          error
        }
      }
    `;

    const rootValue = {};
    const context = getContext({});
    const result = await graphql(schema, mutation, rootValue, context);
    const { data } = result;

    expect(data.registerWithEmail.token).not.toBeNull();
    expect(data.registerWithEmail.error).toBeNull();
  });

  it("should login an user", async () => {
    const user = new User({
      name: "Lana Del Rey",
      email: "delrey@email.com",
      password: "veryGoodPassword"
    });

    await user.save();

    const mutation = `
      mutation {
        loginWithEmail(input: {
        email: "delrey@email.com",
        password: "veryGoodPassword"
      }){
          token
          error
        }
      }
    `;

    const rootValue = {};
    const context = getContext({});
    const result = await graphql(schema, mutation, rootValue, context);
    const { data } = result;

    expect(data.loginWithEmail.token).not.toBeNull();
    expect(data.loginWithEmail.error).toBeNull();
  });

  it("should change the password", async () => {
    const user = new User({
      name: "Lana Del Rey",
      email: "delrey@email.com",
      password: "veryGoodPassword"
    });

    await user.save();

    const mutation = `
      mutation {
        changePassword(input: {
        oldPassword: "veryGoodPassword",
        password: "newVeryGoodPassword"
      }){
          me {
            name
            email
          }
          error
        }
      }
    `;

    const rootValue = {};
    const context = getContext({ user });
    const result = await graphql(schema, mutation, rootValue, context);
    const { data } = result;

    console.log(result);

    expect(data).toMatchSnapshot();
  });
});
