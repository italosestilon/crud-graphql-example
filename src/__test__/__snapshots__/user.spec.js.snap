// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graphql API should change the password 1`] = `
Object {
  "changePassword": Object {
    "error": null,
    "me": Object {
      "email": "delrey@email.com",
      "name": "Lana Del Rey",
    },
  },
}
`;
